### PR TITLE
Add setting DART_DEFINES to flavors page

### PR DESCRIPTION
**Pull request by sfshaza2**
Fixes https://github.com/flutter/website/issues/11178

Please review, @loic-sharma and/or @jmagman 

FYI @sethladd 

**Staged:** https://flutter-docs-prod--pr11192-flavors-qqqcgtun.web.app/deployment/flavors

### CODE REVIEW

File: src/content/deployment/flavors.md [hunk 1]
```diff
@@ -52,32 +58,63 @@ Open your project in Xcode.
...
+and follow these instructions:
+
+</li>
+
+If `DART_DEFINES` doesn't exist yet:
+
+* Click the **+** button in the top left.
+* Select **Add User-Defined**.
+* Enter `DART_DEFINES` as the name.
+* Enter the **base64 encoded version**
```
Comment (loic-sharma) [inline src/content/deployment/flavors.md:71]: Wikipedia seems to always capitalize "Base64", should we as well?

See: https://en.wikipedia.org/wiki/Base64

```suggestion
* Enter the **Base64 encoded version**
```

File: src/content/deployment/flavors.md [hunk 2]
```diff
@@ -52,32 +58,63 @@ Open your project in Xcode.
...
+  of `FLUTTER_APP_FLAVOR=<value>` as the value,
+  where `value` is the name of the flavor
+  that is expected to be the `appFlavor` in Dart.
+
+:::note Important
+You must encode the entire string as `base64`.
+:::
+
+If DART_DEFINES exists, separate the values with a comma.
```
Comment (loic-sharma) [inline src/content/deployment/flavors.md:80]:
```suggestion
If `DART_DEFINES` exists, separate the values with a comma.
```

File: src/content/deployment/flavors.md [hunk 3]
```diff
@@ -52,32 +58,63 @@ Open your project in Xcode.
...
+* Click the **+** button in the top left.
+* Select **Add User-Defined**.
+* Enter `DART_DEFINES` as the name.
+* Enter the **base64 encoded version**
+  of `FLUTTER_APP_FLAVOR=<value>` as the value,
+  where `value` is the name of the flavor
+  that is expected to be the `appFlavor` in Dart.
+
+:::note Important
+You must encode the entire string as `base64`.
```
Comment (loic-sharma) [inline src/content/deployment/flavors.md:77]: If I want my app flavor to be `foo`, should I set `DART_DEFINES` to `FLUTTER_APP_FLAVOR=Zm9v` (only `foo` is encoded) or to `RkxVVFRFUl9BUFBfRkxBVk9SPWZvbw==` (all of `FLUTTER_APP_FLAVOR=foo` is encoded)?

I believe it's the latter, perhaps an example would help clarify?

```suggestion
You must encode the entire string as `base64`.

For example, for flavor `foo` set `DART_DEFINES` to `RkxVVFRFUl9BUFBfRkxBVk9SPWZvbw==` (base64 encoding of `FLUTTER_APP_FLAVOR=foo`).
```

File: src/content/deployment/flavors.md (hunk 2 above)
Comment (loic-sharma) [inline src/content/deployment/flavors.md:80]: Noob question: are you supposed to put the commas before doing the base64 encoding? For example, if I want the Dart defines `HELLO=world` and `FOO=bar`, would my `DART_DEFINES` be `SEVMTE89d29ybGQsRk9PPWJhcg==` (base64 encoding of `HELLO=world,FOO=bar`)?

I think an example would help clarify this case too

File: src/content/deployment/flavors.md [hunk 4]
```diff
@@ -52,32 +58,63 @@ Open your project in Xcode.
...
+
+If `DART_DEFINES` doesn't exist yet:
+
+* Click the **+** button in the top left.
+* Select **Add User-Defined**.
+* Enter `DART_DEFINES` as the name.
+* Enter the **base64 encoded version**
+  of `FLUTTER_APP_FLAVOR=<value>` as the value,
+  where `value` is the name of the flavor
+  that is expected to be the `appFlavor` in Dart.
```
Comment (loic-sharma) [inline src/content/deployment/flavors.md:74]: It looks like this is the first time we reference `appFlavor` in this doc, consider making this a link too:

```suggestion
  that is expected to be the [`appFlavor`][] in Dart.
```

File: src/content/deployment/flavors.md (hunk 2 above)
Comment (sfshaza2) [reply src/content/deployment/flavors.md]: I don't know! That hadn't occurred to me. Who would know in SWE land?

Comment (sfshaza2) [reply src/content/deployment/flavors.md]: @jmagman, do you know?

**Comment by sfshaza2** [comment]
I've added a link to a DartPad where you can encode the text. Can't land this until we know the status of comma delimited strings, encoding-wise.

**Review by jmagman — CHANGES_REQUESTED** [review]
1. I filed https://github.com/flutter/flutter/issues/155951 since this solution seems like a bulky workaround to a feature hole in the Flutter tool (I didn't even know `FLUTTER_APP_FLAVOR` existed--we don't explicitly set it in our iOS flavors samples), and it would be very easy to get out-of-sync with the Xcode configurations.  


2. If we do use this as a workaround, it should use the "free" flavor example on the rest of this page.  These steps need to happen after the flavor is already set up in Xcode, not at the beginning.  https://flutter-docs-prod--pr11192-flavors-qqqcgtun.web.app/deployment/flavors#using-flavors-in-ios-and-macos near the `Product Bundle Identifier` instructions would be better.

3. Over-riding the DART_DEFINES being passed in from the tool in the Xcode configuration is also going to get users into a wonky state.  How are users supposed to know what else the tool is passing into DART_DEFINES? Do they need to know to update it in Xcode if it changes?

> If DART_DEFINES is already defined in Generated.xcconfig, delete that file and enter the DART_DEFINES setting in Xcode before trying again.

I'm not sure what bug is happening in this step in the reported issue https://github.com/flutter/website/issues/11178 but it's concerning, I've never seen this advice before.

**Comment by sfshaza2** [comment]
OK, given @jmagman's comments and the bug she's filed against the s/w, I'm closing this PR, as it can create new problems for a dev. Thanks, @navaronbracke, for surfacing a bug that we didn't know about!

**Pull request closed without merging**